csharp/src/Ice/Reference.cs

```diff
@@ -609,27 +512,83 @@ public class FixedReference : Reference
         // _hashCode is a cached hash code initialized lazily to a value other than 0.
         private int _hashCode = 0;
         private Connection _fixedConnection;
+        private ConnectionRequestHandler _requestHandler;
+
+        internal override Reference Clone()
+        {
+            var clone = (base.Clone() as FixedReference)!;
+            clone._hashCode = 0;
+            return clone;
+        }
+
+        internal override bool IsRequestHandlerCached => true;
 
         internal FixedReference(Communicator communicator,
-                              Identity identity,
-                              string facet,
-                              InvocationMode mode,
-                              bool secure,
-                              Protocol protocol,
-                              Encoding encoding,
-                              Connection connection,
-                              int invocationTimeout,
-                              IReadOnlyDictionary<string, string> context,
-                              bool? compress)
-        : base(communicator, identity, facet, mode, secure, protocol, encoding, invocationTimeout, context)
+                                Identity identity,
+                                string facet,
+                                InvocationMode mode,
+                                bool secure,
+                                Protocol protocol,
+                                Encoding encoding,
+                                Connection connection,
+                                int invocationTimeout,
+                                IReadOnlyDictionary<string, string> context,
+                                bool? compress)
+            : base(communicator, identity, facet, mode, secure, protocol, encoding, invocationTimeout, context)
         {
             _fixedConnection = connection;
-            IsRequestHandlerCached = true;
+
             if (compress is bool compressValue)
             {
                 OverrideCompress = true;
                 Compress = compressValue;
             }
+            _requestHandler = CreateConnectionRequestHandler();
+        }
+
+        // Creates the ConnectionRequestHandler after various checks
+        private ConnectionRequestHandler CreateConnectionRequestHandler()
+        {
+            if (GetMode() == InvocationMode.Datagram)
+            {
+                if (!(_fixedConnection.Endpoint as Endpoint)!.Datagram())
```

Comment:
It's not clear why the cast is needed here.

Reply:
I think it's because `.Endpoint` returns an `IEndpoint` and `Datagram()` is on the `Endpoint` class, not the base interface.